tools/f5map/lib/f5map/app.rb

```diff
@@ -0,0 +1,336 @@
+  # @return [Array<String>] Array of iRule names
+  def virtuals(config)
+    ary = config.subsections.select {|s| s.header.starts_with?('ltm virtual')}
+    ary.collect do |s|
+      # section.header e.g. "ltm virtual DMZ_Internet_VS"
+      s.header.split(/\s+/)[2]
+    end
+  end
+
+  ##
+  # Return a list of virtual servers indicating the target pool and the
+  # associated rules.
+  #
+  # @param [BigIParse::Config] config The F5 configuration instance.  `ltm
+  # virtual` sections are extracted to map out the active iRules.
+  #
+  # @return [Hash<Symbol, <String,Array<String>>] Mapping of the virtual server
+  # configuration with keys :virtual, :pool, :rules, :destination and :vip.
+  def map_rules(config)
+    ary = config.subsections.select {|s| s.header.starts_with?('ltm virtual')}
+    ary.collect do |v|
+      dest = destination(v)
+      vip = dest ? reverse_lookup(dest.split(':').first) : false
+      {
+        virtual: header_name(v.header),
+        pool: pool(v),
+        rules: rules(v),
+        persist: persist(v),
+        destination: dest,
+        vip: vip,
+      }
+    end
+  end
+
+  ##
+  # Parse a BigIParse::Section#header string to return the name.
+  #
+  # @return [String]
+  def header_name(header)
+    header.split(/\s+/).last
+  end
+
+  ##
+  # Extract the destination associated with the F5 Virtual Server configuration.
+  # Returns false if the destination cannot be determined. Some virtual servers,
+  # e.g. `DMZ_Internet_VS` contain a VIP of `0.0.0.0:any` which this method will
+  # return as `"0.0.0.0:any"`
+  #
+  # @param [BigIParse::Config::Section] virtual The `ltm virtual` config
+  #   section instance.
+  #
+  # @return [String] the destination associated with the virtual server or false
+  # if the destination could not be determined.  For example
+  def destination(virtual)
+    if md = DESTINATION_RE.match(virtual.content)
+      md[:destination]
+    else
+      false
+    end
+  end
+
+  ##
+  # Extract the pool from a virtual server config object
+  def pool(virtual)
+    re = /pool\s+(?<pool>[^\s]+)/mx
+    if md = re.match(virtual.content)
+      md[:pool]
+    else
+      false
+    end
+  end
+
+  ##
+  # Extract the rules from a virtual server config object
+  def rules(virtual)
+    # There's surely a better way to do this, but I'm trying to keep the regular
+    # expressions simple and clear.
+    re = /rules\s+{(?<rules>.*?)}/mx # Match the whole inner block
+    rule = /\n\s+(?<rule>[^ \n]+)/mx # Match a single rule inside the block
+    if md = re.match(virtual.content)
+      md[:rules].scan(rule).flatten
+    else
+      []
+    end
+  end
+
+  ##
+  # Extract the persistence elements from a virtual server config object
+  def persist(virtual)
+    # Match to end to parse it out
+    re = /persist\s+(.*)/mx
+    rv = []
+    if md = re.match(virtual.content)
+      if ast = Parser.parse(md.to_s.gsub("\n", ' '))
+        rv = ast.value[1].each_with_object([]) do |v, m|
```

Comment:
FYI, I tend to prefer a functional style with explicit dependency injection, versus an OO style relying on the scope of the closure.  This is why you see me using `each_with_object` instead of `each`.

With that said, will change.

Reply:
That's fair. Don't worry about it then.